storage/durability/sequence_number.rs

```diff
@@ -0,0 +1,208 @@
+        bytes.copy_from_slice(&number_bytes)
+    }
+
+    pub fn to_be_bytes(&self) -> [u8; U80::BYTES] {
+        self.number.to_be_bytes()
+    }
+
+    pub fn invert(&self) -> Self {
+        Self { number: U80::MAX - self.number }
+    }
+
+    pub const fn serialised_len() -> usize {
+        U80::BYTES
+    }
+}
+
+impl From<u128> for SequenceNumber {
+    fn from(value: u128) -> Self {
+        Self::new(U80::new(value))
+    }
+}
+
+pub struct AtomicSequenceNumber {
+    epoch: Cell<u64>,
+    seq_sync: AtomicU64,
+    epoch_increment_mutex: Mutex<()>
+}
+
+impl AtomicSequenceNumber {
+    const SEQ_NUMBER_BITS: u8 = 32;
+    const SYNC_BITS: u8 = (64 - Self::SEQ_NUMBER_BITS);
+
+    const UW_SYNC_MASK: u64 = (1 << Self::SYNC_BITS) - 1;
+    const EPOCH_OVERFLOW_INCREMENT: u64 = (1 << Self::SYNC_BITS);
+
+    #[inline]
+    fn epoch_overflow_oudated(epoch: u64, seq_sync: u64) -> bool {
+        let seq_sync_bits = seq_sync >> Self::SEQ_NUMBER_BITS;
+        let epoch_sync_bits = epoch & Self::UW_SYNC_MASK;
+        seq_sync_bits < epoch_sync_bits
+    }
+
+    #[inline]
+    fn compose(epoch: u64, seq_sync: u64) -> SequenceNumber {
+        SequenceNumber::from((((epoch & !Self::UW_SYNC_MASK) as u128) << Self::SEQ_NUMBER_BITS) | seq_sync as u128)
+    }
+
+    pub fn new(epoch: u64, seq: u32) -> Self {
+        let lw_sync = (epoch & Self::UW_SYNC_MASK) << Self::SEQ_NUMBER_BITS;
+        let seq_sync = (seq as u64) | lw_sync;
+        AtomicSequenceNumber {
+            epoch: Cell::new(epoch),
+            seq_sync: AtomicU64::new(seq_sync),
+            epoch_increment_mutex: Mutex::new(())
+        }
+    }
+
+    pub fn load(&self) -> SequenceNumber {
+        let epoch_read = self.epoch.get();
+        let seq_sync_read = self.seq_sync.load(SeqCst);
+        let epoch_overflow_update = if Self::epoch_overflow_oudated(epoch_read, seq_sync_read) { Self::EPOCH_OVERFLOW_INCREMENT } else { 0 };
+        Self::compose(epoch_read + epoch_overflow_update, seq_sync_read)
+    }
+
+    pub fn fetch_increment_sequence(&self) -> SequenceNumber {
+        // We need to be sure the epoch has not changed.
+        let epoch_read = self.epoch.get();
+        let seq_sync_before_increment = self.seq_sync.fetch_add(1, SeqCst);
+        let epoch_overflow_update = if Self::epoch_overflow_oudated(epoch_read, seq_sync_before_increment) { Self::EPOCH_OVERFLOW_INCREMENT } else { 0 };
+        Self::compose(epoch_read + epoch_overflow_update, seq_sync_before_increment)
+    }
+
+    pub fn increment_epoch_fetch(&self) -> SequenceNumber {
+        let lock = self.epoch_increment_mutex.lock().unwrap();
+        let next_epoch = self.epoch.get() + 1;
+        let seq_sync_updated = next_epoch << Self::SEQ_NUMBER_BITS; // & Self::SEQ_SYNC_MASK; is not needed
+        self.seq_sync.store(seq_sync_updated, SeqCst);
+        self.epoch.set(next_epoch);
+        Self::compose(next_epoch, seq_sync_updated)
+    }
```

Comment:
Not going to check this, but I hope it's all correct!! Maybe add a test with that spins up a bunch of threads (you can use scoped threads) that race on updating a sequence number in a loop and make sure you end up with the right sequence number at the end?